arrow-json/src/raw/mod.rs

```diff
@@ -0,0 +1,502 @@
+use arrow_data::ArrayData;
+use arrow_schema::{ArrowError, DataType, SchemaRef};
+use std::io::BufRead;
+
+mod boolean_array;
+mod list_array;
+mod primitive_array;
+mod string_array;
+mod struct_array;
+mod tape;
+
+/// A builder for [`RawReader`]
+pub struct RawReaderBuilder {
+    batch_size: usize,
+
+    schema: SchemaRef,
+}
+
+impl RawReaderBuilder {
+    /// Create a new [`RawReaderBuilder`] with the provided [`SchemaRef`]
+    ///
+    /// This could be obtained using [`infer_json_schema`] if not known
+    ///
+    /// Any columns not present in `schema` will be ignored
+    ///
+    /// [`infer_json_schema`]: crate::reader::infer_json_schema
+    pub fn new(schema: SchemaRef) -> Self {
+        Self {
+            batch_size: 1024,
+            schema,
+        }
+    }
+
+    /// Sets the batch size in rows to read
+    pub fn with_batch_size(self, batch_size: usize) -> Self {
+        Self { batch_size, ..self }
+    }
+
+    /// Create with the provided [`BufRead`]
+    pub fn build<R: BufRead>(self, reader: R) -> Result<RawReader<R>, ArrowError> {
+        Ok(RawReader {
+            reader,
+            decoder: RawDecoder::try_new(self.schema, self.batch_size)?,
+        })
+    }
+}
+
+/// A [`RecordBatchReader`] that reads newline-delimited JSON data with a known schema
+/// directly into the corresponding arrow arrays
+///
+/// This makes it significantly faster than [`Reader`]
+///
+/// Lines consisting solely of ASCII whitespace are ignored
+///
+/// [`Reader`]: crate::reader::Reader
+pub struct RawReader<R> {
+    reader: R,
+    decoder: RawDecoder,
+}
+
+impl<R> std::fmt::Debug for RawReader<R> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RawReader")
+            .field("decoder", &self.decoder)
+            .finish()
+    }
+}
+
+impl<R: BufRead> RawReader<R> {
+    /// Reads the next [`RecordBatch`] returning `Ok(None)` if EOF
+    fn read(&mut self) -> Result<Option<RecordBatch>, ArrowError> {
+        loop {
+            let buf = self.reader.fill_buf()?;
+            if buf.is_empty() {
+                break;
+            }
+            let read = buf.len();
+
+            let decoded = self.decoder.decode(buf)?;
+            self.reader.consume(decoded);
+            if decoded != read {
+                break;
+            }
+        }
+        self.decoder.flush()
+    }
+}
+
+impl<R: BufRead> Iterator for RawReader<R> {
+    type Item = Result<RecordBatch, ArrowError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.read().transpose()
+    }
+}
+
+impl<R: BufRead> RecordBatchReader for RawReader<R> {
+    fn schema(&self) -> SchemaRef {
+        self.decoder.schema.clone()
+    }
+}
+
+/// [`RawDecoder`] provides a low-level interface for reading JSON data from a byte stream
+///
+/// See [`RawReader`] for a higher-level interface
+pub struct RawDecoder {
+    tape: TapeDecoder,
```

Comment:
nit: I found the use of `tape` for a `TapeDecoder` when `tape` was also used for `Tape` confusing

```suggestion
    tape_decoder: TapeDecoder,
```